feat: store content.child_usage_keys in Container search document [FC-0083]

pomegranited opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Adds child usage keys to the Container search document, so that basic info about Container children (count, usage key, block type) can be retrieved from the search index.

This change is a performance improvement for Library Authors, but does not alter their user experience.

## Supporting information

Part of: https://github.com/openedx/frontend-app-authoring/issues/1778
Addresses this comment: https://github.com/openedx/edx-platform/pull/36477/files#r2033737978
Private-ref: [FAL-4139](https://tasks.opencraft.com/browse/FAL-4139)

## Testing instructions

See https://github.com/openedx/frontend-app-authoring/pull/1820

## Deadline

ASAP

pomegranited committed:
feat: add optional Container object argument

to get_container + get_container_children API methods, to avoid
re-fetching this data during search indexing.

pomegranited committed:
feat: store content.child_usage_keys in Container search document

Stores the draft children + published children (if applicable)

* lib_api.get_container does not take a "user" arg
* fetch_customizable_fields_from_container does not need a "user" arg

openedx-webhooks commented:
Thanks for the pull request, @pomegranited!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited committed:
refactor: moves tags_count into LibraryItem

because anything that appears in a library may be tagged.

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/content-search-unit-blocks

rpenido commented on openedx/core/djangoapps/content_libraries/rest_api/containers.py:
```diff
@@ -184,7 +184,8 @@ def get(self, request, container_key: LibraryContainerLocator):
         )
-        child_entities = api.get_container_children(container_key, published)
+        child_entities = api.get_container_children(container_key, published=published)
+        # TODO -- this looks backwards?
```
I didn't get what you meant here. :sweat_smile:

pomegranited replied:
Ah sorry, I didn't mean to commit that comment :) I was confused when I wrote that, but it looks fine on second look.

rpenido reviewed (approved):
LGTM :+1:
Thank you for your work, @pomegranited!

- [x] I tested this using the instructions from the PR
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

navinkarkera reviewed (approved):
@pomegranited Nice work! :+1:
- [x] I tested this: (Verified children usage_keys in index)
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

bradenmacdonald commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -576,12 +578,21 @@ def searchable_doc_for_container(
 
     try:
-        container = lib_api.get_container(container_key)
+        container_obj = lib_api.get_container_from_key(container_key)
```
The docstring for `get_container_from_key` says it's an "Internal method", but in #36504 we added it to `__all__` so it's now also part of the public API.

The thing is, it doesn't make sense as a public API since both `get_container` and `get_container_from_key` take a container key as their first argument. I think we should rename it to `_get_container_from_key`, and keep it as internal (remove it from `__all__`) and create a separate public function called `get_container_obj` which wraps `_get_container_from_key`. This would be much cleaner and clearer.

pomegranited replied:
I ended up being able to remove `get_container_from_key` from the public API without having to add a `get_container_obj` method. I had to add `container_pk` to `ContainerMetadata` (since ContainerLink needed it), but this seems cleaner to me too.

bradenmacdonald commented on cms/lib/xblock/upstream_sync_container.py:
```diff
@@ -54,15 +54,15 @@ def sync_from_upstream_container(
     Basically, this sets the value of "upstream_display_name" on the downstream block.
     """
-    upstream = lib_api.get_container(LibraryContainerLocator.from_string(downstream.upstream), user)
+    upstream = lib_api.get_container(LibraryContainerLocator.from_string(downstream.upstream))
```
Was this even working before?! Thanks for fixing.

pomegranited replied:
Thank `mypy` :) It caught it when I changed `get_container` to require named parameters after the key.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api/containers.py:
```diff
@@ -192,11 +192,18 @@ def get_container_from_key(container_key: LibraryContainerLocator, isDeleted=Fal
-    container = get_container_from_key(container_key)
+    if not container:
+        container = get_container_from_key(container_key)
+    assert container.key == container_key.container_id
```
This is fine to leave as-is, but I'm personally not a huge fan of this API style of taking an ID and optionally a pre-loaded object.

I would suggest instead:

```python
def get_container(
    container_or_key: LibraryContainerLocator | Container,
    /, *,
    include_collections=False,
) -> ContainerMetadata:
    """
    Get metadata about a container (a Section, Subsection, or Unit).
    """
    container = container_or_key if isinstance(container_or_key, Container) else _get_container_from_key(container_or_key)
```

Otherwise, I don't think the messiness it adds to the code from the API consumer side is worth the two tiny database queries it saves.

The `/, *` will help ensure better compatibility with similar changes to the API in the future. And I think it's better to make this change now to bake it into the new Teak API contract, if others agree. Though all the containers APIs are considered unstable and subject to change anyways.

pomegranited replied:
Reverted this, cf [49557f5](https://github.com/openedx/edx-platform/pull/36528/commits/49557f570e45f640188897773fbbaf48f8bc0d04)

pomegranited committed:
refactor: remove get_container_from_key from public API

API users must use get_container and ContainerMetadata.

Related changes:

* made set_library_item_collections take an entity_key string instead of
  a PublishableEntity instance, so we don't need to fetch a Container
  object to call it.
* added `entity_key_from_usage_key` to the xblock API to support ^
* added container_id
* changed ContainerLink.update_or_create to take the container PK
  instead of a Container object, since this is enough.
* added container_pk to ContainerMetadata to support ^
* reverted previous change that added an optional Container param to
  get_container and get_container_children

bradenmacdonald commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -345,13 +347,10 @@ def _collections_for_content_object(object_id: OpaqueKey) -> dict:
-                container.key,
-            )
+            container = lib_api.get_container(object_id, include_collections=True)
+            collections = container.collections
```
Not a change for this PR: but in the future, ~~I think we should just refactor collections to be an independent API like tagging. Right now we have `get_container(object_id, include_collections=True)` and `container.collections` and `set_container_collections(...)` etc. as well as similar things for components. But why not just `get_collections(entity)`, `add_to_collection(entity)`, etc. that works with any library item? I think that would be cleaner and simpler.~~

Edit: never mind, this just seems to be a problem with the REST API. I don't understand why we have `LibraryBlockCollectionsView` and `LibraryContainerCollectionsView` for updating specific item's collections when we have `LibraryCollectionsView.update_items` which does the same thing generically.

bradenmacdonald commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -293,7 +293,7 @@ def __str__(self):
     def update_or_create(
         cls,
-        upstream_container: Container | None,
+        upstream_container: int | None,
```
Is this working? I would think that below on line 322 you need to set `upstream_container_pk` to this value then, not `upstream_container`. But maybe the Django ORM is sorting it out?

pomegranited replied:
Ah sorry, I had that on my list to fix before pushing, but forgot.
I've fixed it, but I would have expected tests to fail without this change?

bradenmacdonald replied:
I remember seeing a similar thing before where assigning an ID to an object field still worked silently. In any case while I'm testing this I'll make sure the `ContainerLink` is getting created.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/tests/test_api.py:
```diff
@@ -546,11 +548,11 @@ def test_set_library_component_collections(self):
-        component = api.get_component_from_usage_key(UsageKey.from_string(self.lib2_problem_block["id"]))
-
+        component_key = UsageKey.from_string(self.lib2_problem_block["id"])
+        entity_key = xblock_api.entity_key_from_usage_key(component_key)
```
Nit: I kinda prefer this as it was, to keep the logic you copied into `entity_key_from_usage_key` contained within openedx-learning. Can't you leave it as it was and just use:

```
        component = api.get_component_from_usage_key(UsageKey.from_string(self.lib2_problem_block["id"]))
        entity_key = component.publishable_entity.key
```
?

pomegranited replied:
Ya, I was uneasy about that too.. don't like duplicating critical logic like that. Reverted with [2e7be75](https://github.com/openedx/edx-platform/pull/36528/commits/2e7be75c8353211dbb9e3ffcae8f49ebdc9acb9b).

pomegranited committed:
revert: remove xblock_api.entity_key_from_usage_key

Use get_component_from_usage_key, and get it from the component instead.

Also fixes nit re ContainerLink.update_or_create

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/content-search-unit-blocks

bradenmacdonald reviewed (approved):
Thanks, looks good! I did some quick tests with the final version to confirm that the search index seems to be getting updated properly and the MFE is using the search index only.
